catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/action/SourceActionProviderImpl.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.impl.action;
+
+import ddf.action.Action;
+import ddf.action.ActionProvider;
+import ddf.action.impl.ActionImpl;
+import ddf.catalog.source.Source;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.boon.json.JsonFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Convert a JSON string to a map of source identifiers to action URLs. */
+public class SourceActionProviderImpl implements ActionProvider {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SourceActionProviderImpl.class);
+
+  private static final String TITLE = "title";
+
+  private static final String URL = "url";
+
+  private Map<String, Action> sourceIdToAction;
+
+  private String actionProviderId;
+
+  public SourceActionProviderImpl(String actionProviderId) {
+    this.actionProviderId = actionProviderId;
+  }
+
+  @Override
+  public synchronized <T> Action getAction(T subject) {
+
+    if (!(subject instanceof Source)) {
+      return null;
+    }
+
+    Source source = (Source) subject;
+
+    return sourceIdToAction.get(source.getId());
+  }
+
+  @Override
+  public String getId() {
+    return actionProviderId;
+  }
+
+  /**
+   *
+   *
+   * <pre>
+   *     {
+   *         "sourceId1": {
+   *             "title": "Title",
+   *             "url": "http://example.com"
+   *         }
+   *     }
+   * </pre>
+   *
+   * @param sourceActionsJson json representing the source actions
+   */
+  public synchronized void setSourceActions(String sourceActionsJson) {
```

Comment:
❓ Did you consider using the constructor instead of a setter for the actions json?

Reply:
Because the admin can change the configuration through the UI, it needed to be a setter.